// @base-ui-components/react@1.0.0-beta.0/cjs/utils/detectBrowser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isWebKit = exports.isIOS = exports.isFirefox = void 0;
var _utils = require("@floating-ui/react/utils");
const nav = getNavigatorData();
const isWebKit = exports.isWebKit = typeof CSS === 'undefined' || !CSS.supports ? false : CSS.supports('-webkit-backdrop-filter:none');
const isIOS = exports.isIOS =
// iPads can claim to be MacIntel
nav.platform === 'MacIntel' && nav.maxTouchPoints > 1 ? true : /iP(hone|ad|od)|iOS/.test(nav.platform);
const isFirefox = exports.isFirefox = typeof navigator !== 'undefined' && /firefox/i.test((0, _utils.getUserAgent)());

// Avoid Chrome DevTools blue warning.
function getNavigatorData() {
  if (typeof navigator === 'undefined') {
    return {
      platform: '',
      maxTouchPoints: -1
    };
  }
  const uaData = navigator.userAgentData;
  if (uaData?.platform) {
    return {
      platform: uaData.platform,
      maxTouchPoints: navigator.maxTouchPoints
    };
  }
  return {
    platform: navigator.platform,
    maxTouchPoints: navigator.maxTouchPoints
  };
}
