// @base-ui-components/react@1.0.0-beta.0/cjs/utils/useEventCallback.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useEventCallback = useEventCallback;
var React = _interopRequireWildcard(require("react"));
var _useLazyRef = require("./useLazyRef");
// https://github.com/mui/material-ui/issues/41190#issuecomment-2040873379
const useInsertionEffect = React[`useInsertionEffect${Math.random().toFixed(1)}`.slice(0, -3)] || (fn => fn());
function useEventCallback(callback) {
  const stable = (0, _useLazyRef.useLazyRef)(createStableCallback).current;
  stable.next = callback;
  useInsertionEffect(stable.effect);
  return stable.trampoline;
}
function createStableCallback() {
  const stable = {
    next: undefined,
    callback: assertNotCalled,
    trampoline: (...args) => stable.callback?.(...args),
    effect: () => {
      stable.callback = stable.next;
    }
  };
  return stable;
}
function assertNotCalled() {
  throw new Error('Cannot call an event handler while rendering.');
}
