// @base-ui-components/react@1.0.0-beta.0/cjs/utils/usePanelResize.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.usePanelResize = usePanelResize;
var React = _interopRequireWildcard(require("react"));
var _owner = require("./owner");
function setAutoSize(panel) {
  const originalHeight = panel.style.height;
  const originalWidth = panel.style.width;
  panel.style.height = 'auto';
  panel.style.width = 'auto';
  return () => {
    panel.style.height = originalHeight;
    panel.style.width = originalWidth;
  };
}

/**
 * Ensures the panel will expand to the correct height when the window is resized.
 * This prevents content from being cut off or the panel not fitting to the content.
 */
function usePanelResize(panelRef, setDimensions, open) {
  React.useEffect(() => {
    const panel = panelRef.current;
    if (!panel || !open || typeof ResizeObserver === 'undefined') {
      return undefined;
    }
    function recalculateSize() {
      if (!panel) {
        return;
      }
      const cleanup = setAutoSize(panel);
      const scrollHeight = panel.scrollHeight;
      const scrollWidth = panel.scrollWidth;
      cleanup();
      setDimensions(prev => {
        if (prev.height !== scrollHeight || prev.width !== scrollWidth) {
          return {
            height: scrollHeight,
            width: scrollWidth
          };
        }
        return prev;
      });
    }
    const observer = new ResizeObserver(() => {
      if (panel.getAnimations().length > 0) {
        return;
      }
      recalculateSize();
    });
    function handleWindowResize() {
      recalculateSize();
    }
    const win = (0, _owner.ownerWindow)(panel);
    win.addEventListener('resize', handleWindowResize);
    observer.observe(panel);
    return () => {
      observer.disconnect();
      win.removeEventListener('resize', handleWindowResize);
    };
  }, [panelRef, setDimensions, open]);
}
