// @base-ui-components/react@1.0.0-beta.0/esm/toast/close/ToastClose.js
'use client';

import * as React from 'react';
import { useToastRootContext } from "../root/ToastRootContext.js";
import { useToastContext } from "../provider/ToastProviderContext.js";
import { useButton } from "../../use-button/useButton.js";
import { useRenderElement } from "../../utils/useRenderElement.js";

/**
 * Closes the toast when clicked.
 * Renders a `<button>` element.
 *
 * Documentation: [Base UI Toast](https://base-ui.com/react/components/toast)
 */
export const ToastClose = /*#__PURE__*/React.forwardRef(function ToastClose(componentProps, forwardedRef) {
  const {
    render,
    className,
    disabled,
    ...elementProps
  } = componentProps;
  const {
    close
  } = useToastContext();
  const {
    toast
  } = useToastRootContext();
  const {
    getButtonProps
  } = useButton({
    disabled,
    buttonRef: forwardedRef
  });
  const state = React.useMemo(() => ({
    type: toast.type
  }), [toast.type]);
  const element = useRenderElement('button', componentProps, {
    ref: forwardedRef,
    state,
    props: [{
      onClick() {
        close(toast.id);
      }
    }, elementProps, getButtonProps]
  });
  return element;
});
if (process.env.NODE_ENV !== "production") ToastClose.displayName = "ToastClose";
