// @base-ui-components/react@1.0.0-beta.0/esm/toolbar/group/ToolbarGroup.js
'use client';

import * as React from 'react';
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { useToolbarRootContext } from "../root/ToolbarRootContext.js";
import { ToolbarGroupContext } from "./ToolbarGroupContext.js";

/**
 * Groups several toolbar items or toggles.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Toolbar](https://base-ui.com/react/components/toolbar)
 */
import { jsx as _jsx } from "react/jsx-runtime";
export const ToolbarGroup = /*#__PURE__*/React.forwardRef(function ToolbarGroup(props, forwardedRef) {
  const {
    className,
    disabled: disabledProp = false,
    render,
    ...otherProps
  } = props;
  const {
    orientation,
    disabled: toolbarDisabled
  } = useToolbarRootContext();
  const disabled = toolbarDisabled || disabledProp;
  const contextValue = React.useMemo(() => ({
    disabled
  }), [disabled]);
  const state = React.useMemo(() => ({
    disabled,
    orientation
  }), [disabled, orientation]);
  const {
    renderElement
  } = useComponentRenderer({
    render: render ?? 'div',
    ref: forwardedRef,
    state,
    className,
    extraProps: {
      ...otherProps,
      role: 'group'
    }
  });
  return /*#__PURE__*/_jsx(ToolbarGroupContext.Provider, {
    value: contextValue,
    children: renderElement()
  });
});
if (process.env.NODE_ENV !== "production") ToolbarGroup.displayName = "ToolbarGroup";
