// @base-ui-components/react@1.0.0-beta.0/esm/scroll-area/content/ScrollAreaContent.js
'use client';

import * as React from 'react';
import { useModernLayoutEffect } from "../../utils/index.js";
import { useScrollAreaViewportContext } from "../viewport/ScrollAreaViewportContext.js";
import { useRenderElement } from "../../utils/useRenderElement.js";

/**
 * A container for the content of the scroll area.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Scroll Area](https://base-ui.com/react/components/scroll-area)
 */
export const ScrollAreaContent = /*#__PURE__*/React.forwardRef(function ScrollAreaContent(componentProps, forwardedRef) {
  const {
    render,
    className,
    ...elementProps
  } = componentProps;
  const contentWrapperRef = React.useRef(null);
  const {
    computeThumbPosition
  } = useScrollAreaViewportContext();
  useModernLayoutEffect(() => {
    if (typeof ResizeObserver === 'undefined') {
      return undefined;
    }
    const ro = new ResizeObserver(computeThumbPosition);
    if (contentWrapperRef.current) {
      ro.observe(contentWrapperRef.current);
    }
    return () => {
      ro.disconnect();
    };
  }, [computeThumbPosition]);
  const element = useRenderElement('div', componentProps, {
    ref: [forwardedRef, contentWrapperRef],
    props: [{
      role: 'presentation',
      style: {
        minWidth: 'fit-content'
      }
    }, elementProps]
  });
  return element;
});
if (process.env.NODE_ENV !== "production") ScrollAreaContent.displayName = "ScrollAreaContent";
