// @base-ui-components/react@1.0.0-beta.0/esm/menu/popup/MenuPopup.js
'use client';

import * as React from 'react';
import { FloatingFocusManager, useFloatingTree } from '@floating-ui/react';
import { useMenuRootContext } from "../root/MenuRootContext.js";
import { useMenuPositionerContext } from "../positioner/MenuPositionerContext.js";
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { useForkRef } from "../../utils/useForkRef.js";
import { popupStateMapping as baseMapping } from "../../utils/popupStateMapping.js";
import { mergeProps } from "../../merge-props/index.js";
import { transitionStatusMapping } from "../../utils/styleHookMapping.js";
import { useOpenChangeComplete } from "../../utils/useOpenChangeComplete.js";
import { jsx as _jsx } from "react/jsx-runtime";
const customStyleHookMapping = {
  ...baseMapping,
  ...transitionStatusMapping
};
const DISABLED_TRANSITIONS_STYLE = {
  style: {
    transition: 'none'
  }
};
const EMPTY_OBJ = {};

/**
 * A container for the menu items.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Menu](https://base-ui.com/react/components/menu)
 */
export const MenuPopup = /*#__PURE__*/React.forwardRef(function MenuPopup(props, forwardedRef) {
  const {
    render,
    className,
    finalFocus,
    ...other
  } = props;
  const {
    open,
    setOpen,
    popupRef,
    transitionStatus,
    popupProps,
    mounted,
    instantType,
    onOpenChangeComplete,
    parent,
    lastOpenChangeReason
  } = useMenuRootContext();
  const {
    side,
    align,
    floatingContext
  } = useMenuPositionerContext();
  useOpenChangeComplete({
    open,
    ref: popupRef,
    onComplete() {
      if (open) {
        onOpenChangeComplete?.(true);
      }
    }
  });
  const {
    events: menuEvents
  } = useFloatingTree();
  React.useEffect(() => {
    function handleClose(event) {
      setOpen(false, event.domEvent, event.reason);
    }
    menuEvents.on('close', handleClose);
    return () => {
      menuEvents.off('close', handleClose);
    };
  }, [menuEvents, setOpen]);
  const mergedRef = useForkRef(forwardedRef, popupRef);
  const state = React.useMemo(() => ({
    transitionStatus,
    side,
    align,
    open,
    nested: parent.type === 'menu',
    instant: instantType
  }), [transitionStatus, side, align, open, parent.type, instantType]);
  const {
    renderElement
  } = useComponentRenderer({
    render: render || 'div',
    className,
    state,
    extraProps: mergeProps(transitionStatus === 'starting' ? DISABLED_TRANSITIONS_STYLE : EMPTY_OBJ, popupProps, other),
    customStyleHookMapping,
    ref: mergedRef
  });
  let returnFocus = parent.type === undefined || parent.type === 'context-menu';
  if (parent.type === 'menubar' && lastOpenChangeReason !== 'outside-press') {
    returnFocus = true;
  }
  return /*#__PURE__*/_jsx(FloatingFocusManager, {
    context: floatingContext,
    modal: false,
    disabled: !mounted,
    returnFocus: finalFocus || returnFocus,
    initialFocus: parent.type === 'menu' ? -1 : 0,
    restoreFocus: true,
    children: renderElement()
  });
});
if (process.env.NODE_ENV !== "production") MenuPopup.displayName = "MenuPopup";
