// @base-ui-components/react@1.0.0-beta.0/cjs/toast/close/ToastClose.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToastClose = void 0;
var React = _interopRequireWildcard(require("react"));
var _ToastRootContext = require("../root/ToastRootContext");
var _ToastProviderContext = require("../provider/ToastProviderContext");
var _useButton = require("../../use-button/useButton");
var _useRenderElement = require("../../utils/useRenderElement");
/**
 * Closes the toast when clicked.
 * Renders a `<button>` element.
 *
 * Documentation: [Base UI Toast](https://base-ui.com/react/components/toast)
 */
const ToastClose = exports.ToastClose = /*#__PURE__*/React.forwardRef(function ToastClose(componentProps, forwardedRef) {
  const {
    render,
    className,
    disabled,
    ...elementProps
  } = componentProps;
  const {
    close
  } = (0, _ToastProviderContext.useToastContext)();
  const {
    toast
  } = (0, _ToastRootContext.useToastRootContext)();
  const {
    getButtonProps
  } = (0, _useButton.useButton)({
    disabled,
    buttonRef: forwardedRef
  });
  const state = React.useMemo(() => ({
    type: toast.type
  }), [toast.type]);
  const element = (0, _useRenderElement.useRenderElement)('button', componentProps, {
    ref: forwardedRef,
    state,
    props: [{
      onClick() {
        close(toast.id);
      }
    }, elementProps, getButtonProps]
  });
  return element;
});
if (process.env.NODE_ENV !== "production") ToastClose.displayName = "ToastClose";
