// @base-ui-components/react@1.0.0-beta.0/cjs/preview-card/trigger/PreviewCardTrigger.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreviewCardTrigger = void 0;
var React = _interopRequireWildcard(require("react"));
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _PreviewCardContext = require("../root/PreviewCardContext");
var _useForkRef = require("../../utils/useForkRef");
var _popupStateMapping = require("../../utils/popupStateMapping");
/**
 * A link that opens the preview card.
 * Renders an `<a>` element.
 *
 * Documentation: [Base UI Preview Card](https://base-ui.com/react/components/preview-card)
 */
const PreviewCardTrigger = exports.PreviewCardTrigger = /*#__PURE__*/React.forwardRef(function PreviewCardTrigger(props, forwardedRef) {
  const {
    render,
    className,
    ...otherProps
  } = props;
  const {
    open,
    getRootTriggerProps,
    setTriggerElement
  } = (0, _PreviewCardContext.usePreviewCardRootContext)();
  const state = React.useMemo(() => ({
    open
  }), [open]);
  const mergedRef = (0, _useForkRef.useForkRef)(setTriggerElement, forwardedRef);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    propGetter: getRootTriggerProps,
    render: render ?? 'a',
    className,
    state,
    ref: mergedRef,
    extraProps: otherProps,
    customStyleHookMapping: _popupStateMapping.triggerOpenStateMapping
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") PreviewCardTrigger.displayName = "PreviewCardTrigger";
