// @base-ui-components/react@1.0.0-beta.0/cjs/meter/root/MeterRoot.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MeterRoot = void 0;
var React = _interopRequireWildcard(require("react"));
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _MeterRootContext = require("./MeterRootContext");
var _mergeProps = require("../../merge-props");
var _formatNumber = require("../../utils/formatNumber");
var _useLatestRef = require("../../utils/useLatestRef");
var _valueToPercent = require("../../utils/valueToPercent");
var _jsxRuntime = require("react/jsx-runtime");
const EMPTY = {};
function formatValue(value, locale, format) {
  if (!format) {
    return (0, _formatNumber.formatNumber)(value / 100, locale, {
      style: 'percent'
    });
  }
  return (0, _formatNumber.formatNumber)(value, locale, format);
}
/**
 * Groups all parts of the meter and provides the value for screen readers.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Meter](https://base-ui.com/react/components/meter)
 */
const MeterRoot = exports.MeterRoot = /*#__PURE__*/React.forwardRef(function MeterRoot(props, forwardedRef) {
  const {
    format,
    getAriaValueText,
    locale,
    max = 100,
    min = 0,
    value,
    render,
    className,
    ...otherProps
  } = props;
  const formatOptionsRef = (0, _useLatestRef.useLatestRef)(format);
  const [labelId, setLabelId] = React.useState();
  const percentageValue = (0, _valueToPercent.valueToPercent)(value, min, max);
  const formattedValue = formatValue(value, locale, formatOptionsRef.current);
  const propGetter = React.useCallback((externalProps = {}) => {
    let ariaValuetext = `${percentageValue}%`;
    if (getAriaValueText) {
      ariaValuetext = getAriaValueText(formattedValue, value);
    } else if (format) {
      ariaValuetext = formattedValue;
    }
    return (0, _mergeProps.mergeProps)({
      'aria-labelledby': labelId,
      'aria-valuemax': max,
      'aria-valuemin': min,
      'aria-valuenow': percentageValue / 100,
      'aria-valuetext': ariaValuetext,
      role: 'meter'
    }, externalProps);
  }, [format, formattedValue, getAriaValueText, labelId, max, min, value, percentageValue]);
  const contextValue = React.useMemo(() => ({
    formattedValue,
    max,
    min,
    percentageValue,
    setLabelId,
    value
  }), [formattedValue, max, min, percentageValue, setLabelId, value]);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    propGetter,
    render: render ?? 'div',
    state: EMPTY,
    className,
    ref: forwardedRef,
    extraProps: otherProps
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_MeterRootContext.MeterRootContext.Provider, {
    value: contextValue,
    children: renderElement()
  });
});
if (process.env.NODE_ENV !== "production") MeterRoot.displayName = "MeterRoot";
