// @base-ui-components/react@1.0.0-beta.0/cjs/menu/backdrop/MenuBackdrop.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MenuBackdrop = void 0;
var React = _interopRequireWildcard(require("react"));
var _MenuRootContext = require("../root/MenuRootContext");
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _popupStateMapping = require("../../utils/popupStateMapping");
var _styleHookMapping = require("../../utils/styleHookMapping");
var _mergeProps = require("../../merge-props");
var _ContextMenuRootContext = require("../../context-menu/root/ContextMenuRootContext");
const customStyleHookMapping = {
  ..._popupStateMapping.popupStateMapping,
  ..._styleHookMapping.transitionStatusMapping
};

/**
 * An overlay displayed beneath the menu popup.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Menu](https://base-ui.com/react/components/menu)
 */
const MenuBackdrop = exports.MenuBackdrop = /*#__PURE__*/React.forwardRef(function MenuBackdrop(props, forwardedRef) {
  const {
    className,
    render,
    ...other
  } = props;
  const {
    open,
    mounted,
    transitionStatus,
    lastOpenChangeReason
  } = (0, _MenuRootContext.useMenuRootContext)();
  const contextMenuContext = (0, _ContextMenuRootContext.useContextMenuRootContext)();
  const state = React.useMemo(() => ({
    open,
    transitionStatus
  }), [open, transitionStatus]);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    render: render ?? 'div',
    className,
    state,
    ref: contextMenuContext?.backdropRef ? [forwardedRef, contextMenuContext.backdropRef] : forwardedRef,
    extraProps: (0, _mergeProps.mergeProps)({
      role: 'presentation',
      hidden: !mounted,
      style: {
        pointerEvents: lastOpenChangeReason === 'trigger-hover' ? 'none' : undefined,
        userSelect: 'none',
        WebkitUserSelect: 'none'
      }
    }, other),
    customStyleHookMapping
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") MenuBackdrop.displayName = "MenuBackdrop";
