// @base-ui-components/react@1.0.0-beta.0/esm/menu/backdrop/MenuBackdrop.js
'use client';

import * as React from 'react';
import { useMenuRootContext } from "../root/MenuRootContext.js";
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { popupStateMapping as baseMapping } from "../../utils/popupStateMapping.js";
import { transitionStatusMapping } from "../../utils/styleHookMapping.js";
import { mergeProps } from "../../merge-props/index.js";
import { useContextMenuRootContext } from "../../context-menu/root/ContextMenuRootContext.js";
const customStyleHookMapping = {
  ...baseMapping,
  ...transitionStatusMapping
};

/**
 * An overlay displayed beneath the menu popup.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Menu](https://base-ui.com/react/components/menu)
 */
export const MenuBackdrop = /*#__PURE__*/React.forwardRef(function MenuBackdrop(props, forwardedRef) {
  const {
    className,
    render,
    ...other
  } = props;
  const {
    open,
    mounted,
    transitionStatus,
    lastOpenChangeReason
  } = useMenuRootContext();
  const contextMenuContext = useContextMenuRootContext();
  const state = React.useMemo(() => ({
    open,
    transitionStatus
  }), [open, transitionStatus]);
  const {
    renderElement
  } = useComponentRenderer({
    render: render ?? 'div',
    className,
    state,
    ref: contextMenuContext?.backdropRef ? [forwardedRef, contextMenuContext.backdropRef] : forwardedRef,
    extraProps: mergeProps({
      role: 'presentation',
      hidden: !mounted,
      style: {
        pointerEvents: lastOpenChangeReason === 'trigger-hover' ? 'none' : undefined,
        userSelect: 'none',
        WebkitUserSelect: 'none'
      }
    }, other),
    customStyleHookMapping
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") MenuBackdrop.displayName = "MenuBackdrop";
