// @base-ui-components/react@1.0.0-beta.0/cjs/checkbox/indicator/CheckboxIndicator.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CheckboxIndicator = void 0;
var React = _interopRequireWildcard(require("react"));
var _CheckboxRootContext = require("../root/CheckboxRootContext");
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _useCustomStyleHookMapping = require("../utils/useCustomStyleHookMapping");
var _useOpenChangeComplete = require("../../utils/useOpenChangeComplete");
var _useTransitionStatus = require("../../utils/useTransitionStatus");
var _useForkRef = require("../../utils/useForkRef");
var _styleHookMapping = require("../../utils/styleHookMapping");
var _constants = require("../../field/utils/constants");
/**
 * Indicates whether the checkbox is ticked.
 * Renders a `<span>` element.
 *
 * Documentation: [Base UI Checkbox](https://base-ui.com/react/components/checkbox)
 */
const CheckboxIndicator = exports.CheckboxIndicator = /*#__PURE__*/React.forwardRef(function CheckboxIndicator(props, forwardedRef) {
  const {
    render,
    className,
    keepMounted = false,
    ...otherProps
  } = props;
  const rootState = (0, _CheckboxRootContext.useCheckboxRootContext)();
  const rendered = rootState.checked || rootState.indeterminate;
  const {
    transitionStatus,
    setMounted
  } = (0, _useTransitionStatus.useTransitionStatus)(rendered);
  const indicatorRef = React.useRef(null);
  const mergedRef = (0, _useForkRef.useForkRef)(forwardedRef, indicatorRef);
  const state = React.useMemo(() => ({
    ...rootState,
    transitionStatus
  }), [rootState, transitionStatus]);
  (0, _useOpenChangeComplete.useOpenChangeComplete)({
    open: rendered,
    ref: indicatorRef,
    onComplete() {
      if (!rendered) {
        setMounted(false);
      }
    }
  });
  const baseStyleHookMapping = (0, _useCustomStyleHookMapping.useCustomStyleHookMapping)(rootState);
  const customStyleHookMapping = React.useMemo(() => ({
    ...baseStyleHookMapping,
    ..._styleHookMapping.transitionStatusMapping,
    ..._constants.fieldValidityMapping
  }), [baseStyleHookMapping]);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    render: render ?? 'span',
    ref: mergedRef,
    state,
    className,
    customStyleHookMapping,
    extraProps: otherProps
  });
  const shouldRender = keepMounted || rendered;
  if (!shouldRender) {
    return null;
  }
  return renderElement();
});
if (process.env.NODE_ENV !== "production") CheckboxIndicator.displayName = "CheckboxIndicator";
