// @base-ui-components/react@1.0.0-beta.0/esm/toggle/Toggle.js
'use client';

import * as React from 'react';
import { useControlled } from "../utils/useControlled.js";
import { useEventCallback } from "../utils/useEventCallback.js";
import { useRenderElement } from "../utils/useRenderElement.js";
import { CompositeItem } from "../composite/item/CompositeItem.js";
import { useToggleGroupContext } from "../toggle-group/ToggleGroupContext.js";
import { useButton } from "../use-button/useButton.js";

/**
 * A two-state button that can be on or off.
 * Renders a `<button>` element.
 *
 * Documentation: [Base UI Toggle](https://base-ui.com/react/components/toggle)
 */
import { jsx as _jsx } from "react/jsx-runtime";
export const Toggle = /*#__PURE__*/React.forwardRef(function Toggle(componentProps, forwardedRef) {
  const {
    className,
    defaultPressed: defaultPressedProp = false,
    disabled: disabledProp = false,
    form,
    // never participates in form validation
    onPressedChange: onPressedChangeProp,
    pressed: pressedProp,
    render,
    type,
    // cannot change button type
    value: valueProp,
    ...elementProps
  } = componentProps;
  const value = valueProp ?? '';
  const groupContext = useToggleGroupContext();
  const groupValue = groupContext?.value ?? [];
  const defaultPressed = groupContext ? undefined : defaultPressedProp;
  const disabled = (disabledProp || groupContext?.disabled) ?? false;
  const [pressed, setPressedState] = useControlled({
    controlled: groupContext && value ? groupValue?.indexOf(value) > -1 : pressedProp,
    default: defaultPressed,
    name: 'Toggle',
    state: 'pressed'
  });
  const onPressedChange = useEventCallback((nextPressed, event) => {
    groupContext?.setGroupValue?.(value, nextPressed, event);
    onPressedChangeProp?.(nextPressed, event);
  });
  const {
    getButtonProps,
    buttonRef
  } = useButton({
    disabled,
    buttonRef: forwardedRef
  });
  const state = React.useMemo(() => ({
    disabled,
    pressed
  }), [disabled, pressed]);
  const element = useRenderElement('button', componentProps, {
    state,
    ref: buttonRef,
    props: [{
      'aria-pressed': pressed,
      onClick(event) {
        const nextPressed = !pressed;
        setPressedState(nextPressed);
        onPressedChange(nextPressed, event.nativeEvent);
      },
      ref: buttonRef
    }, elementProps, getButtonProps]
  });
  return groupContext ? /*#__PURE__*/_jsx(CompositeItem, {
    render: element
  }) : element;
});
if (process.env.NODE_ENV !== "production") Toggle.displayName = "Toggle";
