// @base-ui-components/react@1.0.0-beta.0/cjs/tooltip/index.parts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Arrow", {
  enumerable: true,
  get: function () {
    return _TooltipArrow.TooltipArrow;
  }
});
Object.defineProperty(exports, "Popup", {
  enumerable: true,
  get: function () {
    return _TooltipPopup.TooltipPopup;
  }
});
Object.defineProperty(exports, "Portal", {
  enumerable: true,
  get: function () {
    return _TooltipPortal.TooltipPortal;
  }
});
Object.defineProperty(exports, "Positioner", {
  enumerable: true,
  get: function () {
    return _TooltipPositioner.TooltipPositioner;
  }
});
Object.defineProperty(exports, "Provider", {
  enumerable: true,
  get: function () {
    return _TooltipProvider.TooltipProvider;
  }
});
Object.defineProperty(exports, "Root", {
  enumerable: true,
  get: function () {
    return _TooltipRoot.TooltipRoot;
  }
});
Object.defineProperty(exports, "Trigger", {
  enumerable: true,
  get: function () {
    return _TooltipTrigger.TooltipTrigger;
  }
});
var _TooltipRoot = require("./root/TooltipRoot");
var _TooltipTrigger = require("./trigger/TooltipTrigger");
var _TooltipPortal = require("./portal/TooltipPortal");
var _TooltipPositioner = require("./positioner/TooltipPositioner");
var _TooltipPopup = require("./popup/TooltipPopup");
var _TooltipArrow = require("./arrow/TooltipArrow");
var _TooltipProvider = require("./provider/TooltipProvider");
