// @base-ui-components/react@1.0.0-beta.0/cjs/menu/trigger/MenuTrigger.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MenuTrigger = void 0;
var React = _interopRequireWildcard(require("react"));
var _utils = require("@floating-ui/react/utils");
var _CompositeItem = require("../../composite/item/CompositeItem");
var _MenuRootContext = require("../root/MenuRootContext");
var _popupStateMapping = require("../../utils/popupStateMapping");
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _useForkRef = require("../../utils/useForkRef");
var _mergeProps = require("../../merge-props");
var _useButton = require("../../use-button/useButton");
var _useTimeout = require("../../utils/useTimeout");
var _owner = require("../../utils/owner");
var _getPseudoElementBounds = require("../../utils/getPseudoElementBounds");
var _useEventCallback = require("../../utils/useEventCallback");
var _jsxRuntime = require("react/jsx-runtime");
const BOUNDARY_OFFSET = 2;

/**
 * A button that opens the menu.
 * Renders a `<button>` element.
 *
 * Documentation: [Base UI Menu](https://base-ui.com/react/components/menu)
 */
const MenuTrigger = exports.MenuTrigger = /*#__PURE__*/React.forwardRef(function MenuTrigger(props, forwardedRef) {
  const {
    render,
    className,
    disabled: disabledProp = false,
    ...other
  } = props;
  const {
    triggerProps: rootTriggerProps,
    disabled: menuDisabled,
    setTriggerElement,
    open,
    setOpen,
    allowMouseUpTriggerRef,
    positionerRef,
    parent,
    lastOpenChangeReason
  } = (0, _MenuRootContext.useMenuRootContext)();
  const disabled = disabledProp || menuDisabled;
  const triggerRef = React.useRef(null);
  const mergedRef = (0, _useForkRef.useForkRef)(forwardedRef, triggerRef);
  const allowMouseUpTriggerTimeout = (0, _useTimeout.useTimeout)();
  const {
    getButtonProps,
    buttonRef
  } = (0, _useButton.useButton)({
    disabled,
    buttonRef: mergedRef
  });
  const handleRef = (0, _useForkRef.useForkRef)(buttonRef, setTriggerElement);
  React.useEffect(() => {
    if (!open && parent.type === undefined) {
      allowMouseUpTriggerRef.current = false;
    }
  }, [allowMouseUpTriggerRef, open, parent.type]);
  const handleMouseUp = (0, _useEventCallback.useEventCallback)(mouseEvent => {
    if (!triggerRef.current) {
      return;
    }
    allowMouseUpTriggerTimeout.clear();
    allowMouseUpTriggerRef.current = false;
    const mouseUpTarget = mouseEvent.target;
    if ((0, _utils.contains)(triggerRef.current, mouseUpTarget) || (0, _utils.contains)(positionerRef.current, mouseUpTarget) || mouseUpTarget === triggerRef.current) {
      return;
    }
    const bounds = (0, _getPseudoElementBounds.getPseudoElementBounds)(triggerRef.current);
    if (mouseEvent.clientX >= bounds.left - BOUNDARY_OFFSET && mouseEvent.clientX <= bounds.right + BOUNDARY_OFFSET && mouseEvent.clientY >= bounds.top - BOUNDARY_OFFSET && mouseEvent.clientY <= bounds.bottom + BOUNDARY_OFFSET) {
      return;
    }
    setOpen(false, mouseEvent, 'cancel-open');
  });
  React.useEffect(() => {
    if (open && lastOpenChangeReason === 'trigger-hover') {
      const doc = (0, _owner.ownerDocument)(triggerRef.current);
      doc.addEventListener('mouseup', handleMouseUp, {
        once: true
      });
    }
  }, [open, handleMouseUp, lastOpenChangeReason]);
  const getTriggerProps = React.useCallback(externalProps => {
    return (0, _mergeProps.mergeProps)({
      'aria-haspopup': 'menu',
      ref: handleRef,
      onMouseDown: event => {
        if (open) {
          return;
        }

        // mousedown -> mouseup on menu item should not trigger it within 200ms.
        allowMouseUpTriggerTimeout.start(200, () => {
          allowMouseUpTriggerRef.current = true;
        });
        const doc = (0, _owner.ownerDocument)(event.currentTarget);
        doc.addEventListener('mouseup', handleMouseUp, {
          once: true
        });
      }
    }, externalProps, getButtonProps);
  }, [getButtonProps, handleRef, open, allowMouseUpTriggerRef, allowMouseUpTriggerTimeout, handleMouseUp]);
  const state = React.useMemo(() => ({
    disabled,
    open
  }), [disabled, open]);
  const propGetter = React.useCallback(externalProps => (0, _mergeProps.mergeProps)(rootTriggerProps, externalProps, getTriggerProps), [getTriggerProps, rootTriggerProps]);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    render: render || 'button',
    className,
    state,
    propGetter,
    customStyleHookMapping: _popupStateMapping.pressableTriggerOpenStateMapping,
    extraProps: other
  });
  if (parent.type === 'menubar') {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_CompositeItem.CompositeItem, {
      render: renderElement()
    });
  }
  return renderElement();
});
if (process.env.NODE_ENV !== "production") MenuTrigger.displayName = "MenuTrigger";
