// @base-ui-components/react@1.0.0-beta.0/cjs/meter/label/MeterLabel.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MeterLabel = void 0;
var React = _interopRequireWildcard(require("react"));
var _mergeProps = require("../../merge-props");
var _useBaseUiId = require("../../utils/useBaseUiId");
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _useModernLayoutEffect = require("../../utils/useModernLayoutEffect");
var _MeterRootContext = require("../root/MeterRootContext");
const EMPTY = {};
/**
 * An accessible label for the meter.
 * Renders a `<span>` element.
 *
 * Documentation: [Base UI Meter](https://base-ui.com/react/components/meter)
 */
const MeterLabel = exports.MeterLabel = /*#__PURE__*/React.forwardRef(function MeterLabel(props, forwardedRef) {
  const {
    render,
    className,
    id: idProp,
    ...otherProps
  } = props;
  const id = (0, _useBaseUiId.useBaseUiId)(idProp);
  const {
    setLabelId
  } = (0, _MeterRootContext.useMeterRootContext)();
  (0, _useModernLayoutEffect.useModernLayoutEffect)(() => {
    setLabelId(id);
    return () => setLabelId(undefined);
  }, [id, setLabelId]);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    render: render ?? 'span',
    state: EMPTY,
    className,
    ref: forwardedRef,
    extraProps: (0, _mergeProps.mergeProps)({
      id
    }, otherProps)
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") MeterLabel.displayName = "MeterLabel";
