// @base-ui-components/react@1.0.0-beta.0/cjs/navigation-menu/list/NavigationMenuList.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NavigationMenuList = void 0;
var React = _interopRequireWildcard(require("react"));
var _useRenderElement = require("../../utils/useRenderElement");
var _CompositeRoot = require("../../composite/root/CompositeRoot");
var _NavigationMenuRootContext = require("../root/NavigationMenuRootContext");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * Contains a list of navigation menu items.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Navigation Menu](https://base-ui.com/react/components/navigation-menu)
 */
const NavigationMenuList = exports.NavigationMenuList = /*#__PURE__*/React.forwardRef(function NavigationMenuList(componentProps, forwardedRef) {
  const {
    className,
    render,
    ...elementProps
  } = componentProps;
  const {
    orientation,
    open
  } = (0, _NavigationMenuRootContext.useNavigationMenuRootContext)();
  const state = React.useMemo(() => ({
    open
  }), [open]);
  const element = (0, _useRenderElement.useRenderElement)('div', componentProps, {
    state,
    ref: forwardedRef,
    props: elementProps
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_CompositeRoot.CompositeRoot, {
    render: element,
    loop: false,
    orientation: orientation,
    stopEventPropagation: true
  });
});
if (process.env.NODE_ENV !== "production") NavigationMenuList.displayName = "NavigationMenuList";
