// @tanstack/react-form@1.14.1/dist/cjs/createFormHook.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("react/jsx-runtime");
const react = require("react");
const useForm = require("./useForm.cjs");
function createFormHookContexts() {
  const fieldContext = react.createContext(null);
  function useFieldContext() {
    const field = react.useContext(fieldContext);
    if (!field) {
      throw new Error(
        "`fieldContext` only works when within a `fieldComponent` passed to `createFormHook`"
      );
    }
    return field;
  }
  const formContext = react.createContext(null);
  function useFormContext() {
    const form = react.useContext(formContext);
    if (!form) {
      throw new Error(
        "`formContext` only works when within a `formComponent` passed to `createFormHook`"
      );
    }
    return form;
  }
  return { fieldContext, useFieldContext, useFormContext, formContext };
}
function createFormHook({
  fieldComponents,
  fieldContext,
  formContext,
  formComponents
}) {
  function useAppForm(props) {
    const form = useForm.useForm(props);
    const AppForm = react.useMemo(() => {
      const AppForm2 = ({ children }) => {
        return /* @__PURE__ */ jsxRuntime.jsx(formContext.Provider, { value: form, children });
      };
      return AppForm2;
    }, [form]);
    const AppField = react.useMemo(() => {
      const AppField2 = ({ children, ...props2 }) => {
        return /* @__PURE__ */ jsxRuntime.jsx(form.Field, { ...props2, children: (field) => (
          // eslint-disable-next-line @eslint-react/no-context-provider
          /* @__PURE__ */ jsxRuntime.jsx(fieldContext.Provider, { value: field, children: children(Object.assign(field, fieldComponents)) })
        ) });
      };
      return AppField2;
    }, [form]);
    const extendedForm = react.useMemo(() => {
      return Object.assign(form, {
        AppField,
        AppForm,
        ...formComponents
      });
    }, [form, AppField, AppForm]);
    return extendedForm;
  }
  function withForm({
    render,
    props
  }) {
    return (innerProps) => render({ ...props, ...innerProps });
  }
  return {
    useAppForm,
    withForm
  };
}
exports.createFormHook = createFormHook;
exports.createFormHookContexts = createFormHookContexts;
