// @base-ui-components/react@1.0.0-beta.0/esm/avatar/image/AvatarImage.js
'use client';

import * as React from 'react';
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { useEventCallback } from "../../utils/useEventCallback.js";
import { useModernLayoutEffect } from "../../utils/useModernLayoutEffect.js";
import { useAvatarRootContext } from "../root/AvatarRootContext.js";
import { avatarStyleHookMapping } from "../root/styleHooks.js";
import { useImageLoadingStatus } from "./useImageLoadingStatus.js";

/**
 * The image to be displayed in the avatar.
 * Renders an `<img>` element.
 *
 * Documentation: [Base UI Avatar](https://base-ui.com/react/components/avatar)
 */
export const AvatarImage = /*#__PURE__*/React.forwardRef(function AvatarImage(props, forwardedRef) {
  const {
    className,
    render,
    onLoadingStatusChange: onLoadingStatusChangeProp,
    referrerPolicy,
    crossOrigin,
    ...otherProps
  } = props;
  const context = useAvatarRootContext();
  const imageLoadingStatus = useImageLoadingStatus(props.src, {
    referrerPolicy,
    crossOrigin
  });
  const handleLoadingStatusChange = useEventCallback(status => {
    onLoadingStatusChangeProp?.(status);
    context.setImageLoadingStatus(status);
  });
  useModernLayoutEffect(() => {
    if (imageLoadingStatus !== 'idle') {
      handleLoadingStatusChange(imageLoadingStatus);
    }
  }, [imageLoadingStatus, handleLoadingStatusChange]);
  const state = React.useMemo(() => ({
    imageLoadingStatus
  }), [imageLoadingStatus]);
  const {
    renderElement
  } = useComponentRenderer({
    render: render ?? 'img',
    state,
    className,
    ref: forwardedRef,
    extraProps: otherProps,
    customStyleHookMapping: avatarStyleHookMapping
  });
  return imageLoadingStatus === 'loaded' ? renderElement() : null;
});
if (process.env.NODE_ENV !== "production") AvatarImage.displayName = "AvatarImage";
