// @base-ui-components/react@1.0.0-beta.0/esm/dialog/title/DialogTitle.js
'use client';

import * as React from 'react';
import { useDialogRootContext } from "../root/DialogRootContext.js";
import { mergeProps } from "../../merge-props/index.js";
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { useModernLayoutEffect } from "../../utils/useModernLayoutEffect.js";
import { useBaseUiId } from "../../utils/useBaseUiId.js";
const state = {};

/**
 * A heading that labels the dialog.
 * Renders an `<h2>` element.
 *
 * Documentation: [Base UI Dialog](https://base-ui.com/react/components/dialog)
 */
export const DialogTitle = /*#__PURE__*/React.forwardRef(function DialogTitle(props, forwardedRef) {
  const {
    render,
    className,
    id: idProp,
    ...other
  } = props;
  const {
    setTitleElementId
  } = useDialogRootContext();
  const id = useBaseUiId(idProp);
  useModernLayoutEffect(() => {
    setTitleElementId(id);
    return () => {
      setTitleElementId(undefined);
    };
  }, [id, setTitleElementId]);
  const getProps = React.useCallback((externalProps = {}) => mergeProps({
    id
  }, externalProps), [id]);
  const {
    renderElement
  } = useComponentRenderer({
    propGetter: getProps,
    render: render ?? 'h2',
    className,
    state,
    ref: forwardedRef,
    extraProps: other
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") DialogTitle.displayName = "DialogTitle";
