// @base-ui-components/react@1.0.0-beta.0/cjs/dialog/trigger/DialogTrigger.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DialogTrigger = void 0;
var React = _interopRequireWildcard(require("react"));
var _DialogRootContext = require("../root/DialogRootContext");
var _useButton = require("../../use-button/useButton");
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _useForkRef = require("../../utils/useForkRef");
var _popupStateMapping = require("../../utils/popupStateMapping");
/**
 * A button that opens the dialog.
 * Renders a `<button>` element.
 *
 * Documentation: [Base UI Dialog](https://base-ui.com/react/components/dialog)
 */
const DialogTrigger = exports.DialogTrigger = /*#__PURE__*/React.forwardRef(function DialogTrigger(props, forwardedRef) {
  const {
    render,
    className,
    disabled = false,
    ...other
  } = props;
  const {
    open,
    setTriggerElement,
    getTriggerProps
  } = (0, _DialogRootContext.useDialogRootContext)();
  const state = React.useMemo(() => ({
    disabled,
    open
  }), [disabled, open]);
  const mergedRef = (0, _useForkRef.useForkRef)(forwardedRef, setTriggerElement);
  const {
    getButtonProps
  } = (0, _useButton.useButton)({
    disabled,
    buttonRef: mergedRef
  });
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    render: render ?? 'button',
    className,
    state,
    propGetter: externalProps => getButtonProps(getTriggerProps(externalProps)),
    extraProps: other,
    customStyleHookMapping: _popupStateMapping.triggerOpenStateMapping
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") DialogTrigger.displayName = "DialogTrigger";
