// @base-ui-components/react@1.0.0-beta.0/esm/menu/arrow/MenuArrow.js
'use client';

import * as React from 'react';
import { useMenuPositionerContext } from "../positioner/MenuPositionerContext.js";
import { useMenuRootContext } from "../root/MenuRootContext.js";
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { useForkRef } from "../../utils/useForkRef.js";
import { popupStateMapping } from "../../utils/popupStateMapping.js";

/**
 * Displays an element positioned against the menu anchor.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Menu](https://base-ui.com/react/components/menu)
 */
export const MenuArrow = /*#__PURE__*/React.forwardRef(function MenuArrow(props, forwardedRef) {
  const {
    className,
    render,
    ...otherProps
  } = props;
  const {
    open
  } = useMenuRootContext();
  const {
    arrowRef,
    side,
    align,
    arrowUncentered,
    arrowStyles
  } = useMenuPositionerContext();
  const state = React.useMemo(() => ({
    open,
    side,
    align,
    uncentered: arrowUncentered
  }), [open, side, align, arrowUncentered]);
  const mergedRef = useForkRef(arrowRef, forwardedRef);
  const {
    renderElement
  } = useComponentRenderer({
    render: render ?? 'div',
    className,
    state,
    ref: mergedRef,
    extraProps: {
      style: arrowStyles,
      'aria-hidden': true,
      ...otherProps
    },
    customStyleHookMapping: popupStateMapping
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") MenuArrow.displayName = "MenuArrow";
