// @base-ui-components/react@1.0.0-beta.0/esm/fieldset/legend/FieldsetLegend.js
'use client';

import * as React from 'react';
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { useFieldsetLegend } from "./useFieldsetLegend.js";
import { useFieldsetRootContext } from "../root/FieldsetRootContext.js";
/**
 * An accessible label that is automatically associated with the fieldset.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Fieldset](https://base-ui.com/react/components/fieldset)
 */
export const FieldsetLegend = /*#__PURE__*/React.forwardRef(function FieldsetLegend(props, forwardedRef) {
  const {
    render,
    className,
    id,
    ...otherProps
  } = props;
  const {
    getLegendProps
  } = useFieldsetLegend({
    id
  });
  const {
    disabled
  } = useFieldsetRootContext();
  const state = React.useMemo(() => ({
    disabled: disabled ?? false
  }), [disabled]);
  const {
    renderElement
  } = useComponentRenderer({
    propGetter: getLegendProps,
    ref: forwardedRef,
    render: render ?? 'div',
    className,
    state,
    extraProps: otherProps
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") FieldsetLegend.displayName = "FieldsetLegend";
