// @base-ui-components/react@1.0.0-beta.0/esm/alert-dialog/description/AlertDialogDescription.js
'use client';

import * as React from 'react';
import { useAlertDialogRootContext } from "../root/AlertDialogRootContext.js";
import { mergeProps } from "../../merge-props/index.js";
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { useModernLayoutEffect } from "../../utils/useModernLayoutEffect.js";
import { useBaseUiId } from "../../utils/useBaseUiId.js";
const state = {};

/**
 * A paragraph with additional information about the alert dialog.
 * Renders a `<p>` element.
 *
 * Documentation: [Base UI Alert Dialog](https://base-ui.com/react/components/alert-dialog)
 */
export const AlertDialogDescription = /*#__PURE__*/React.forwardRef(function AlertDialogDescription(props, forwardedRef) {
  const {
    render,
    className,
    id: idProp,
    ...other
  } = props;
  const {
    setDescriptionElementId
  } = useAlertDialogRootContext();
  const id = useBaseUiId(idProp);
  useModernLayoutEffect(() => {
    setDescriptionElementId(id);
    return () => {
      setDescriptionElementId(undefined);
    };
  }, [id, setDescriptionElementId]);
  const getProps = React.useCallback((externalProps = {}) => mergeProps({
    id
  }, externalProps), [id]);
  const {
    renderElement
  } = useComponentRenderer({
    propGetter: getProps,
    render: render ?? 'p',
    className,
    state,
    ref: forwardedRef,
    extraProps: other
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") AlertDialogDescription.displayName = "AlertDialogDescription";
