// @base-ui-components/react@1.0.0-beta.0/cjs/preview-card/arrow/PreviewCardArrow.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreviewCardArrow = void 0;
var React = _interopRequireWildcard(require("react"));
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _PreviewCardPositionerContext = require("../positioner/PreviewCardPositionerContext");
var _useForkRef = require("../../utils/useForkRef");
var _PreviewCardContext = require("../root/PreviewCardContext");
var _popupStateMapping = require("../../utils/popupStateMapping");
var _mergeProps = require("../../merge-props");
/**
 * Displays an element positioned against the preview card anchor.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Preview Card](https://base-ui.com/react/components/preview-card)
 */
const PreviewCardArrow = exports.PreviewCardArrow = /*#__PURE__*/React.forwardRef(function PreviewCardArrow(props, forwardedRef) {
  const {
    render,
    className,
    ...otherProps
  } = props;
  const {
    open
  } = (0, _PreviewCardContext.usePreviewCardRootContext)();
  const {
    arrowRef,
    side,
    align,
    arrowUncentered,
    arrowStyles
  } = (0, _PreviewCardPositionerContext.usePreviewCardPositionerContext)();
  const getArrowProps = React.useCallback((externalProps = {}) => (0, _mergeProps.mergeProps)({
    style: arrowStyles,
    'aria-hidden': true
  }, externalProps), [arrowStyles]);
  const state = React.useMemo(() => ({
    open,
    side,
    align,
    uncentered: arrowUncentered
  }), [open, side, align, arrowUncentered]);
  const mergedRef = (0, _useForkRef.useForkRef)(arrowRef, forwardedRef);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    propGetter: getArrowProps,
    render: render ?? 'div',
    className,
    state,
    ref: mergedRef,
    extraProps: otherProps,
    customStyleHookMapping: _popupStateMapping.popupStateMapping
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") PreviewCardArrow.displayName = "PreviewCardArrow";
