// @base-ui-components/react@1.0.0-beta.0/esm/checkbox/indicator/CheckboxIndicator.js
'use client';

import * as React from 'react';
import { useCheckboxRootContext } from "../root/CheckboxRootContext.js";
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { useCustomStyleHookMapping } from "../utils/useCustomStyleHookMapping.js";
import { useOpenChangeComplete } from "../../utils/useOpenChangeComplete.js";
import { useTransitionStatus } from "../../utils/useTransitionStatus.js";
import { useForkRef } from "../../utils/useForkRef.js";
import { transitionStatusMapping } from "../../utils/styleHookMapping.js";
import { fieldValidityMapping } from "../../field/utils/constants.js";

/**
 * Indicates whether the checkbox is ticked.
 * Renders a `<span>` element.
 *
 * Documentation: [Base UI Checkbox](https://base-ui.com/react/components/checkbox)
 */
export const CheckboxIndicator = /*#__PURE__*/React.forwardRef(function CheckboxIndicator(props, forwardedRef) {
  const {
    render,
    className,
    keepMounted = false,
    ...otherProps
  } = props;
  const rootState = useCheckboxRootContext();
  const rendered = rootState.checked || rootState.indeterminate;
  const {
    transitionStatus,
    setMounted
  } = useTransitionStatus(rendered);
  const indicatorRef = React.useRef(null);
  const mergedRef = useForkRef(forwardedRef, indicatorRef);
  const state = React.useMemo(() => ({
    ...rootState,
    transitionStatus
  }), [rootState, transitionStatus]);
  useOpenChangeComplete({
    open: rendered,
    ref: indicatorRef,
    onComplete() {
      if (!rendered) {
        setMounted(false);
      }
    }
  });
  const baseStyleHookMapping = useCustomStyleHookMapping(rootState);
  const customStyleHookMapping = React.useMemo(() => ({
    ...baseStyleHookMapping,
    ...transitionStatusMapping,
    ...fieldValidityMapping
  }), [baseStyleHookMapping]);
  const {
    renderElement
  } = useComponentRenderer({
    render: render ?? 'span',
    ref: mergedRef,
    state,
    className,
    customStyleHookMapping,
    extraProps: otherProps
  });
  const shouldRender = keepMounted || rendered;
  if (!shouldRender) {
    return null;
  }
  return renderElement();
});
if (process.env.NODE_ENV !== "production") CheckboxIndicator.displayName = "CheckboxIndicator";
