// @base-ui-components/react@1.0.0-beta.0/esm/menu/group-label/MenuGroupLabel.js
'use client';

import * as React from 'react';
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { useBaseUiId } from "../../utils/useBaseUiId.js";
import { useMenuGroupRootContext } from "../group/MenuGroupContext.js";
import { useModernLayoutEffect } from "../../utils/useModernLayoutEffect.js";
const state = {};

/**
 * An accessible label that is automatically associated with its parent group.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Menu](https://base-ui.com/react/components/menu)
 */
export const MenuGroupLabel = /*#__PURE__*/React.forwardRef(function MenuGroupLabelComponent(props, forwardedRef) {
  const {
    className,
    render,
    id: idProp,
    ...other
  } = props;
  const id = useBaseUiId(idProp);
  const {
    setLabelId
  } = useMenuGroupRootContext();
  useModernLayoutEffect(() => {
    setLabelId(id);
    return () => {
      setLabelId(undefined);
    };
  }, [setLabelId, id]);
  const {
    renderElement
  } = useComponentRenderer({
    render: render ?? 'div',
    className,
    state,
    extraProps: {
      role: 'presentation',
      id,
      ...other
    },
    ref: forwardedRef
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") MenuGroupLabel.displayName = "MenuGroupLabel";
