// @base-ui-components/react@1.0.0-beta.0/esm/alert-dialog/trigger/AlertDialogTrigger.js
'use client';

import * as React from 'react';
import { useAlertDialogRootContext } from "../root/AlertDialogRootContext.js";
import { useButton } from "../../use-button/useButton.js";
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { useForkRef } from "../../utils/useForkRef.js";
import { triggerOpenStateMapping } from "../../utils/popupStateMapping.js";

/**
 * A button that opens the alert dialog.
 * Renders a `<button>` element.
 *
 * Documentation: [Base UI Alert Dialog](https://base-ui.com/react/components/alert-dialog)
 */
export const AlertDialogTrigger = /*#__PURE__*/React.forwardRef(function AlertDialogTrigger(props, forwardedRef) {
  const {
    render,
    className,
    disabled = false,
    ...other
  } = props;
  const {
    open,
    setTriggerElement,
    getTriggerProps
  } = useAlertDialogRootContext();
  const state = React.useMemo(() => ({
    disabled,
    open
  }), [disabled, open]);
  const mergedRef = useForkRef(forwardedRef, setTriggerElement);
  const {
    getButtonProps
  } = useButton({
    disabled,
    buttonRef: mergedRef
  });
  const {
    renderElement
  } = useComponentRenderer({
    render: render ?? 'button',
    className,
    state,
    propGetter: externalProps => getButtonProps(getTriggerProps(externalProps)),
    extraProps: other,
    customStyleHookMapping: triggerOpenStateMapping,
    ref: mergedRef
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") AlertDialogTrigger.displayName = "AlertDialogTrigger";
