// @base-ui-components/react@1.0.0-beta.0/esm/preview-card/arrow/PreviewCardArrow.js
'use client';

import * as React from 'react';
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { usePreviewCardPositionerContext } from "../positioner/PreviewCardPositionerContext.js";
import { useForkRef } from "../../utils/useForkRef.js";
import { usePreviewCardRootContext } from "../root/PreviewCardContext.js";
import { popupStateMapping } from "../../utils/popupStateMapping.js";
import { mergeProps } from "../../merge-props/index.js";

/**
 * Displays an element positioned against the preview card anchor.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Preview Card](https://base-ui.com/react/components/preview-card)
 */
export const PreviewCardArrow = /*#__PURE__*/React.forwardRef(function PreviewCardArrow(props, forwardedRef) {
  const {
    render,
    className,
    ...otherProps
  } = props;
  const {
    open
  } = usePreviewCardRootContext();
  const {
    arrowRef,
    side,
    align,
    arrowUncentered,
    arrowStyles
  } = usePreviewCardPositionerContext();
  const getArrowProps = React.useCallback((externalProps = {}) => mergeProps({
    style: arrowStyles,
    'aria-hidden': true
  }, externalProps), [arrowStyles]);
  const state = React.useMemo(() => ({
    open,
    side,
    align,
    uncentered: arrowUncentered
  }), [open, side, align, arrowUncentered]);
  const mergedRef = useForkRef(arrowRef, forwardedRef);
  const {
    renderElement
  } = useComponentRenderer({
    propGetter: getArrowProps,
    render: render ?? 'div',
    className,
    state,
    ref: mergedRef,
    extraProps: otherProps,
    customStyleHookMapping: popupStateMapping
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") PreviewCardArrow.displayName = "PreviewCardArrow";
