// @base-ui-components/react@1.0.0-beta.0/cjs/alert-dialog/title/AlertDialogTitle.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AlertDialogTitle = void 0;
var React = _interopRequireWildcard(require("react"));
var _AlertDialogRootContext = require("../root/AlertDialogRootContext");
var _mergeProps = require("../../merge-props");
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _useModernLayoutEffect = require("../../utils/useModernLayoutEffect");
var _useBaseUiId = require("../../utils/useBaseUiId");
const state = {};

/**
 * A heading that labels the dialog.
 * Renders an `<h2>` element.
 *
 * Documentation: [Base UI Alert Dialog](https://base-ui.com/react/components/alert-dialog)
 */
const AlertDialogTitle = exports.AlertDialogTitle = /*#__PURE__*/React.forwardRef(function AlertDialogTitle(props, forwardedRef) {
  const {
    render,
    className,
    id: idProp,
    ...other
  } = props;
  const {
    setTitleElementId
  } = (0, _AlertDialogRootContext.useAlertDialogRootContext)();
  const id = (0, _useBaseUiId.useBaseUiId)(idProp);
  (0, _useModernLayoutEffect.useModernLayoutEffect)(() => {
    setTitleElementId(id);
    return () => {
      setTitleElementId(undefined);
    };
  }, [id, setTitleElementId]);
  const getProps = React.useCallback((externalProps = {}) => (0, _mergeProps.mergeProps)({
    id
  }, externalProps), [id]);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    propGetter: getProps,
    render: render ?? 'h2',
    className,
    state,
    ref: forwardedRef,
    extraProps: other
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") AlertDialogTitle.displayName = "AlertDialogTitle";
