// @base-ui-components/react@1.0.0-beta.0/cjs/switch/thumb/SwitchThumb.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SwitchThumb = void 0;
var React = _interopRequireWildcard(require("react"));
var _SwitchRootContext = require("../root/SwitchRootContext");
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _FieldRootContext = require("../../field/root/FieldRootContext");
var _styleHooks = require("../styleHooks");
/**
 * The movable part of the switch that indicates whether the switch is on or off.
 * Renders a `<span>`.
 *
 * Documentation: [Base UI Switch](https://base-ui.com/react/components/switch)
 */
const SwitchThumb = exports.SwitchThumb = /*#__PURE__*/React.forwardRef(function SwitchThumb(props, forwardedRef) {
  const {
    render,
    className,
    ...other
  } = props;
  const {
    state: fieldState
  } = (0, _FieldRootContext.useFieldRootContext)();
  const state = (0, _SwitchRootContext.useSwitchRootContext)();
  const extendedState = {
    ...fieldState,
    ...state
  };
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    render: render || 'span',
    className,
    state: extendedState,
    extraProps: other,
    customStyleHookMapping: _styleHooks.styleHookMapping,
    ref: forwardedRef
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") SwitchThumb.displayName = "SwitchThumb";
