// @base-ui-components/react@1.0.0-beta.0/esm/meter/label/MeterLabel.js
'use client';

import * as React from 'react';
import { mergeProps } from "../../merge-props/index.js";
import { useBaseUiId } from "../../utils/useBaseUiId.js";
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { useModernLayoutEffect } from "../../utils/useModernLayoutEffect.js";
import { useMeterRootContext } from "../root/MeterRootContext.js";
const EMPTY = {};
/**
 * An accessible label for the meter.
 * Renders a `<span>` element.
 *
 * Documentation: [Base UI Meter](https://base-ui.com/react/components/meter)
 */
export const MeterLabel = /*#__PURE__*/React.forwardRef(function MeterLabel(props, forwardedRef) {
  const {
    render,
    className,
    id: idProp,
    ...otherProps
  } = props;
  const id = useBaseUiId(idProp);
  const {
    setLabelId
  } = useMeterRootContext();
  useModernLayoutEffect(() => {
    setLabelId(id);
    return () => setLabelId(undefined);
  }, [id, setLabelId]);
  const {
    renderElement
  } = useComponentRenderer({
    render: render ?? 'span',
    state: EMPTY,
    className,
    ref: forwardedRef,
    extraProps: mergeProps({
      id
    }, otherProps)
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") MeterLabel.displayName = "MeterLabel";
