// @base-ui-components/react@1.0.0-beta.0/cjs/fieldset/legend/FieldsetLegend.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FieldsetLegend = void 0;
var React = _interopRequireWildcard(require("react"));
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _useFieldsetLegend = require("./useFieldsetLegend");
var _FieldsetRootContext = require("../root/FieldsetRootContext");
/**
 * An accessible label that is automatically associated with the fieldset.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Fieldset](https://base-ui.com/react/components/fieldset)
 */
const FieldsetLegend = exports.FieldsetLegend = /*#__PURE__*/React.forwardRef(function FieldsetLegend(props, forwardedRef) {
  const {
    render,
    className,
    id,
    ...otherProps
  } = props;
  const {
    getLegendProps
  } = (0, _useFieldsetLegend.useFieldsetLegend)({
    id
  });
  const {
    disabled
  } = (0, _FieldsetRootContext.useFieldsetRootContext)();
  const state = React.useMemo(() => ({
    disabled: disabled ?? false
  }), [disabled]);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    propGetter: getLegendProps,
    ref: forwardedRef,
    render: render ?? 'div',
    className,
    state,
    extraProps: otherProps
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") FieldsetLegend.displayName = "FieldsetLegend";
