// @base-ui-components/react@1.0.0-beta.0/esm/preview-card/root/PreviewCardRoot.js
'use client';

import * as React from 'react';
import { PreviewCardRootContext } from "./PreviewCardContext.js";
import { usePreviewCardRoot } from "./usePreviewCardRoot.js";
import { CLOSE_DELAY, OPEN_DELAY } from "../utils/constants.js";
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * Groups all parts of the preview card.
 * Doesn’t render its own HTML element.
 *
 * Documentation: [Base UI Preview Card](https://base-ui.com/react/components/preview-card)
 */
export const PreviewCardRoot = function PreviewCardRoot(props) {
  const {
    delay,
    closeDelay,
    onOpenChangeComplete,
    actionsRef
  } = props;
  const delayWithDefault = delay ?? OPEN_DELAY;
  const closeDelayWithDefault = closeDelay ?? CLOSE_DELAY;
  const previewCardRoot = usePreviewCardRoot({
    delay,
    closeDelay,
    actionsRef,
    onOpenChangeComplete,
    open: props.open,
    onOpenChange: props.onOpenChange,
    defaultOpen: props.defaultOpen
  });
  const contextValue = React.useMemo(() => ({
    ...previewCardRoot,
    delay: delayWithDefault,
    closeDelay: closeDelayWithDefault
  }), [closeDelayWithDefault, delayWithDefault, previewCardRoot]);
  return /*#__PURE__*/_jsx(PreviewCardRootContext.Provider, {
    value: contextValue,
    children: props.children
  });
};
if (process.env.NODE_ENV !== "production") PreviewCardRoot.displayName = "PreviewCardRoot";
