// @base-ui-components/react@1.0.0-beta.0/esm/menu/group/MenuGroup.js
'use client';

import * as React from 'react';
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { MenuGroupContext } from "./MenuGroupContext.js";
import { jsx as _jsx } from "react/jsx-runtime";
const state = {};

/**
 * Groups related menu items with the corresponding label.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Menu](https://base-ui.com/react/components/menu)
 */
export const MenuGroup = /*#__PURE__*/React.forwardRef(function MenuGroup(props, forwardedRef) {
  const {
    render,
    className,
    ...other
  } = props;
  const [labelId, setLabelId] = React.useState(undefined);
  const context = React.useMemo(() => ({
    setLabelId
  }), [setLabelId]);
  const {
    renderElement
  } = useComponentRenderer({
    render: render || 'div',
    className,
    state,
    extraProps: {
      role: 'group',
      'aria-labelledby': labelId,
      ...other
    },
    ref: forwardedRef
  });
  return /*#__PURE__*/_jsx(MenuGroupContext.Provider, {
    value: context,
    children: renderElement()
  });
});
if (process.env.NODE_ENV !== "production") MenuGroup.displayName = "MenuGroup";
