// @base-ui-components/react@1.0.0-beta.0/cjs/toolbar/group/ToolbarGroup.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToolbarGroup = void 0;
var React = _interopRequireWildcard(require("react"));
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _ToolbarRootContext = require("../root/ToolbarRootContext");
var _ToolbarGroupContext = require("./ToolbarGroupContext");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * Groups several toolbar items or toggles.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Toolbar](https://base-ui.com/react/components/toolbar)
 */
const ToolbarGroup = exports.ToolbarGroup = /*#__PURE__*/React.forwardRef(function ToolbarGroup(props, forwardedRef) {
  const {
    className,
    disabled: disabledProp = false,
    render,
    ...otherProps
  } = props;
  const {
    orientation,
    disabled: toolbarDisabled
  } = (0, _ToolbarRootContext.useToolbarRootContext)();
  const disabled = toolbarDisabled || disabledProp;
  const contextValue = React.useMemo(() => ({
    disabled
  }), [disabled]);
  const state = React.useMemo(() => ({
    disabled,
    orientation
  }), [disabled, orientation]);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    render: render ?? 'div',
    ref: forwardedRef,
    state,
    className,
    extraProps: {
      ...otherProps,
      role: 'group'
    }
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_ToolbarGroupContext.ToolbarGroupContext.Provider, {
    value: contextValue,
    children: renderElement()
  });
});
if (process.env.NODE_ENV !== "production") ToolbarGroup.displayName = "ToolbarGroup";
