// @base-ui-components/react@1.0.0-beta.0/cjs/preview-card/root/PreviewCardRoot.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreviewCardRoot = void 0;
var React = _interopRequireWildcard(require("react"));
var _PreviewCardContext = require("./PreviewCardContext");
var _usePreviewCardRoot = require("./usePreviewCardRoot");
var _constants = require("../utils/constants");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * Groups all parts of the preview card.
 * Doesn’t render its own HTML element.
 *
 * Documentation: [Base UI Preview Card](https://base-ui.com/react/components/preview-card)
 */
const PreviewCardRoot = exports.PreviewCardRoot = function PreviewCardRoot(props) {
  const {
    delay,
    closeDelay,
    onOpenChangeComplete,
    actionsRef
  } = props;
  const delayWithDefault = delay ?? _constants.OPEN_DELAY;
  const closeDelayWithDefault = closeDelay ?? _constants.CLOSE_DELAY;
  const previewCardRoot = (0, _usePreviewCardRoot.usePreviewCardRoot)({
    delay,
    closeDelay,
    actionsRef,
    onOpenChangeComplete,
    open: props.open,
    onOpenChange: props.onOpenChange,
    defaultOpen: props.defaultOpen
  });
  const contextValue = React.useMemo(() => ({
    ...previewCardRoot,
    delay: delayWithDefault,
    closeDelay: closeDelayWithDefault
  }), [closeDelayWithDefault, delayWithDefault, previewCardRoot]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_PreviewCardContext.PreviewCardRootContext.Provider, {
    value: contextValue,
    children: props.children
  });
};
if (process.env.NODE_ENV !== "production") PreviewCardRoot.displayName = "PreviewCardRoot";
