// @base-ui-components/react@1.0.0-beta.0/cjs/menu/popup/MenuPopup.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MenuPopup = void 0;
var React = _interopRequireWildcard(require("react"));
var _react2 = require("@floating-ui/react");
var _MenuRootContext = require("../root/MenuRootContext");
var _MenuPositionerContext = require("../positioner/MenuPositionerContext");
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _useForkRef = require("../../utils/useForkRef");
var _popupStateMapping = require("../../utils/popupStateMapping");
var _mergeProps = require("../../merge-props");
var _styleHookMapping = require("../../utils/styleHookMapping");
var _useOpenChangeComplete = require("../../utils/useOpenChangeComplete");
var _jsxRuntime = require("react/jsx-runtime");
const customStyleHookMapping = {
  ..._popupStateMapping.popupStateMapping,
  ..._styleHookMapping.transitionStatusMapping
};
const DISABLED_TRANSITIONS_STYLE = {
  style: {
    transition: 'none'
  }
};
const EMPTY_OBJ = {};

/**
 * A container for the menu items.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Menu](https://base-ui.com/react/components/menu)
 */
const MenuPopup = exports.MenuPopup = /*#__PURE__*/React.forwardRef(function MenuPopup(props, forwardedRef) {
  const {
    render,
    className,
    finalFocus,
    ...other
  } = props;
  const {
    open,
    setOpen,
    popupRef,
    transitionStatus,
    popupProps,
    mounted,
    instantType,
    onOpenChangeComplete,
    parent,
    lastOpenChangeReason
  } = (0, _MenuRootContext.useMenuRootContext)();
  const {
    side,
    align,
    floatingContext
  } = (0, _MenuPositionerContext.useMenuPositionerContext)();
  (0, _useOpenChangeComplete.useOpenChangeComplete)({
    open,
    ref: popupRef,
    onComplete() {
      if (open) {
        onOpenChangeComplete?.(true);
      }
    }
  });
  const {
    events: menuEvents
  } = (0, _react2.useFloatingTree)();
  React.useEffect(() => {
    function handleClose(event) {
      setOpen(false, event.domEvent, event.reason);
    }
    menuEvents.on('close', handleClose);
    return () => {
      menuEvents.off('close', handleClose);
    };
  }, [menuEvents, setOpen]);
  const mergedRef = (0, _useForkRef.useForkRef)(forwardedRef, popupRef);
  const state = React.useMemo(() => ({
    transitionStatus,
    side,
    align,
    open,
    nested: parent.type === 'menu',
    instant: instantType
  }), [transitionStatus, side, align, open, parent.type, instantType]);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    render: render || 'div',
    className,
    state,
    extraProps: (0, _mergeProps.mergeProps)(transitionStatus === 'starting' ? DISABLED_TRANSITIONS_STYLE : EMPTY_OBJ, popupProps, other),
    customStyleHookMapping,
    ref: mergedRef
  });
  let returnFocus = parent.type === undefined || parent.type === 'context-menu';
  if (parent.type === 'menubar' && lastOpenChangeReason !== 'outside-press') {
    returnFocus = true;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_react2.FloatingFocusManager, {
    context: floatingContext,
    modal: false,
    disabled: !mounted,
    returnFocus: finalFocus || returnFocus,
    initialFocus: parent.type === 'menu' ? -1 : 0,
    restoreFocus: true,
    children: renderElement()
  });
});
if (process.env.NODE_ENV !== "production") MenuPopup.displayName = "MenuPopup";
