// @base-ui-components/react@1.0.0-beta.0/cjs/navigation-menu/content/NavigationMenuContent.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NavigationMenuContent = void 0;
var React = _interopRequireWildcard(require("react"));
var ReactDOM = _interopRequireWildcard(require("react-dom"));
var _react2 = require("@floating-ui/react");
var _utils = require("@floating-ui/react/utils");
var _useRenderElement = require("../../utils/useRenderElement");
var _NavigationMenuRootContext = require("../root/NavigationMenuRootContext");
var _NavigationMenuItemContext = require("../item/NavigationMenuItemContext");
var _useTransitionStatus = require("../../utils/useTransitionStatus");
var _useOpenChangeComplete = require("../../utils/useOpenChangeComplete");
var _styleHookMapping = require("../../utils/styleHookMapping");
var _CompositeRoot = require("../../composite/root/CompositeRoot");
var _inertValue = require("../../utils/inertValue");
var _popupStateMapping = require("../../utils/popupStateMapping");
var _jsxRuntime = require("react/jsx-runtime");
const customStyleHookMapping = {
  ..._popupStateMapping.popupStateMapping,
  ..._styleHookMapping.transitionStatusMapping,
  activationDirection(value) {
    if (value === 'left') {
      return {
        'data-activation-direction': 'left'
      };
    }
    if (value === 'right') {
      return {
        'data-activation-direction': 'right'
      };
    }
    return null;
  }
};

/**
 * A container for the content of the navigation menu item that is moved into the popup
 * when the item is active.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Navigation Menu](https://base-ui.com/react/components/navigation-menu)
 */
const NavigationMenuContent = exports.NavigationMenuContent = /*#__PURE__*/React.forwardRef(function NavigationMenuContent(componentProps, forwardedRef) {
  const {
    className,
    render,
    ...elementProps
  } = componentProps;
  const {
    mounted: popupMounted,
    viewportElement,
    value,
    activationDirection,
    currentContentRef
  } = (0, _NavigationMenuRootContext.useNavigationMenuRootContext)();
  const itemValue = (0, _NavigationMenuItemContext.useNavigationMenuItemContext)();
  const nodeId = (0, _NavigationMenuRootContext.useNavigationMenuTreeContext)();
  const open = popupMounted && value === itemValue;
  const ref = React.useRef(null);
  const [focusInside, setFocusInside] = React.useState(false);
  const {
    mounted,
    setMounted,
    transitionStatus
  } = (0, _useTransitionStatus.useTransitionStatus)(open);
  (0, _useOpenChangeComplete.useOpenChangeComplete)({
    ref,
    open,
    onComplete() {
      if (!open) {
        setMounted(false);
      }
    }
  });
  const state = React.useMemo(() => ({
    open,
    transitionStatus,
    activationDirection
  }), [open, transitionStatus, activationDirection]);
  const handleCurrentContentRef = React.useCallback(node => {
    if (node) {
      currentContentRef.current = node;
    }
  }, [currentContentRef]);
  const commonProps = {
    onFocus() {
      setFocusInside(true);
    },
    onBlur(event) {
      if (!(0, _utils.contains)(event.currentTarget, event.relatedTarget)) {
        setFocusInside(false);
      }
    }
  };
  const shouldRender = viewportElement !== null && mounted;
  const element = (0, _useRenderElement.useRenderElement)('div', componentProps, {
    enabled: shouldRender,
    state,
    ref: [forwardedRef, ref, handleCurrentContentRef],
    props: [!open && mounted ? {
      style: {
        position: 'absolute',
        top: 0,
        left: 0
      },
      inert: (0, _inertValue.inertValue)(!focusInside),
      ...commonProps
    } : commonProps, elementProps],
    customStyleHookMapping
  });
  if (!viewportElement || !element) {
    return null;
  }
  return /*#__PURE__*/ReactDOM.createPortal(/*#__PURE__*/(0, _jsxRuntime.jsx)(_react2.FloatingNode, {
    id: nodeId,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_CompositeRoot.CompositeRoot, {
      render: element,
      stopEventPropagation: true
    })
  }), viewportElement);
});
if (process.env.NODE_ENV !== "production") NavigationMenuContent.displayName = "NavigationMenuContent";
