// @base-ui-components/react@1.0.0-beta.0/cjs/separator/Separator.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Separator = void 0;
var React = _interopRequireWildcard(require("react"));
var _mergeProps = require("../merge-props");
var _useComponentRenderer = require("../utils/useComponentRenderer");
/**
 * A separator element accessible to screen readers.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Separator](https://base-ui.com/react/components/separator)
 */
const Separator = exports.Separator = /*#__PURE__*/React.forwardRef(function SeparatorComponent(props, forwardedRef) {
  const {
    className,
    render,
    orientation = 'horizontal',
    ...other
  } = props;
  const state = React.useMemo(() => ({
    orientation
  }), [orientation]);
  const getSeparatorProps = React.useCallback((externalProps = {}) => (0, _mergeProps.mergeProps)({
    'aria-orientation': orientation
  }, externalProps), [orientation]);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    propGetter: getSeparatorProps,
    render: render ?? 'div',
    className,
    state,
    extraProps: {
      role: 'separator',
      ...other
    },
    ref: forwardedRef
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") Separator.displayName = "Separator";
