// @base-ui-components/react@1.0.0-beta.0/cjs/alert-dialog/index.parts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Backdrop", {
  enumerable: true,
  get: function () {
    return _AlertDialogBackdrop.AlertDialogBackdrop;
  }
});
Object.defineProperty(exports, "Close", {
  enumerable: true,
  get: function () {
    return _AlertDialogClose.AlertDialogClose;
  }
});
Object.defineProperty(exports, "Description", {
  enumerable: true,
  get: function () {
    return _AlertDialogDescription.AlertDialogDescription;
  }
});
Object.defineProperty(exports, "Popup", {
  enumerable: true,
  get: function () {
    return _AlertDialogPopup.AlertDialogPopup;
  }
});
Object.defineProperty(exports, "Portal", {
  enumerable: true,
  get: function () {
    return _AlertDialogPortal.AlertDialogPortal;
  }
});
Object.defineProperty(exports, "Root", {
  enumerable: true,
  get: function () {
    return _AlertDialogRoot.AlertDialogRoot;
  }
});
Object.defineProperty(exports, "Title", {
  enumerable: true,
  get: function () {
    return _AlertDialogTitle.AlertDialogTitle;
  }
});
Object.defineProperty(exports, "Trigger", {
  enumerable: true,
  get: function () {
    return _AlertDialogTrigger.AlertDialogTrigger;
  }
});
var _AlertDialogBackdrop = require("./backdrop/AlertDialogBackdrop");
var _AlertDialogClose = require("./close/AlertDialogClose");
var _AlertDialogDescription = require("./description/AlertDialogDescription");
var _AlertDialogPopup = require("./popup/AlertDialogPopup");
var _AlertDialogPortal = require("./portal/AlertDialogPortal");
var _AlertDialogRoot = require("./root/AlertDialogRoot");
var _AlertDialogTitle = require("./title/AlertDialogTitle");
var _AlertDialogTrigger = require("./trigger/AlertDialogTrigger");
