// @base-ui-components/react@1.0.0-beta.0/esm/switch/thumb/SwitchThumb.js
'use client';

import * as React from 'react';
import { useSwitchRootContext } from "../root/SwitchRootContext.js";
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { useFieldRootContext } from "../../field/root/FieldRootContext.js";
import { styleHookMapping } from "../styleHooks.js";

/**
 * The movable part of the switch that indicates whether the switch is on or off.
 * Renders a `<span>`.
 *
 * Documentation: [Base UI Switch](https://base-ui.com/react/components/switch)
 */
export const SwitchThumb = /*#__PURE__*/React.forwardRef(function SwitchThumb(props, forwardedRef) {
  const {
    render,
    className,
    ...other
  } = props;
  const {
    state: fieldState
  } = useFieldRootContext();
  const state = useSwitchRootContext();
  const extendedState = {
    ...fieldState,
    ...state
  };
  const {
    renderElement
  } = useComponentRenderer({
    render: render || 'span',
    className,
    state: extendedState,
    extraProps: other,
    customStyleHookMapping: styleHookMapping,
    ref: forwardedRef
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") SwitchThumb.displayName = "SwitchThumb";
