// @base-ui-components/react@1.0.0-beta.0/esm/meter/root/MeterRoot.js
'use client';

import * as React from 'react';
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { MeterRootContext } from "./MeterRootContext.js";
import { mergeProps } from "../../merge-props/index.js";
import { formatNumber } from "../../utils/formatNumber.js";
import { useLatestRef } from "../../utils/useLatestRef.js";
import { valueToPercent } from "../../utils/valueToPercent.js";
import { jsx as _jsx } from "react/jsx-runtime";
const EMPTY = {};
function formatValue(value, locale, format) {
  if (!format) {
    return formatNumber(value / 100, locale, {
      style: 'percent'
    });
  }
  return formatNumber(value, locale, format);
}
/**
 * Groups all parts of the meter and provides the value for screen readers.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Meter](https://base-ui.com/react/components/meter)
 */
export const MeterRoot = /*#__PURE__*/React.forwardRef(function MeterRoot(props, forwardedRef) {
  const {
    format,
    getAriaValueText,
    locale,
    max = 100,
    min = 0,
    value,
    render,
    className,
    ...otherProps
  } = props;
  const formatOptionsRef = useLatestRef(format);
  const [labelId, setLabelId] = React.useState();
  const percentageValue = valueToPercent(value, min, max);
  const formattedValue = formatValue(value, locale, formatOptionsRef.current);
  const propGetter = React.useCallback((externalProps = {}) => {
    let ariaValuetext = `${percentageValue}%`;
    if (getAriaValueText) {
      ariaValuetext = getAriaValueText(formattedValue, value);
    } else if (format) {
      ariaValuetext = formattedValue;
    }
    return mergeProps({
      'aria-labelledby': labelId,
      'aria-valuemax': max,
      'aria-valuemin': min,
      'aria-valuenow': percentageValue / 100,
      'aria-valuetext': ariaValuetext,
      role: 'meter'
    }, externalProps);
  }, [format, formattedValue, getAriaValueText, labelId, max, min, value, percentageValue]);
  const contextValue = React.useMemo(() => ({
    formattedValue,
    max,
    min,
    percentageValue,
    setLabelId,
    value
  }), [formattedValue, max, min, percentageValue, setLabelId, value]);
  const {
    renderElement
  } = useComponentRenderer({
    propGetter,
    render: render ?? 'div',
    state: EMPTY,
    className,
    ref: forwardedRef,
    extraProps: otherProps
  });
  return /*#__PURE__*/_jsx(MeterRootContext.Provider, {
    value: contextValue,
    children: renderElement()
  });
});
if (process.env.NODE_ENV !== "production") MeterRoot.displayName = "MeterRoot";
