// @base-ui-components/react@1.0.0-beta.0/cjs/menu/arrow/MenuArrow.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MenuArrow = void 0;
var React = _interopRequireWildcard(require("react"));
var _MenuPositionerContext = require("../positioner/MenuPositionerContext");
var _MenuRootContext = require("../root/MenuRootContext");
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _useForkRef = require("../../utils/useForkRef");
var _popupStateMapping = require("../../utils/popupStateMapping");
/**
 * Displays an element positioned against the menu anchor.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Menu](https://base-ui.com/react/components/menu)
 */
const MenuArrow = exports.MenuArrow = /*#__PURE__*/React.forwardRef(function MenuArrow(props, forwardedRef) {
  const {
    className,
    render,
    ...otherProps
  } = props;
  const {
    open
  } = (0, _MenuRootContext.useMenuRootContext)();
  const {
    arrowRef,
    side,
    align,
    arrowUncentered,
    arrowStyles
  } = (0, _MenuPositionerContext.useMenuPositionerContext)();
  const state = React.useMemo(() => ({
    open,
    side,
    align,
    uncentered: arrowUncentered
  }), [open, side, align, arrowUncentered]);
  const mergedRef = (0, _useForkRef.useForkRef)(arrowRef, forwardedRef);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    render: render ?? 'div',
    className,
    state,
    ref: mergedRef,
    extraProps: {
      style: arrowStyles,
      'aria-hidden': true,
      ...otherProps
    },
    customStyleHookMapping: _popupStateMapping.popupStateMapping
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") MenuArrow.displayName = "MenuArrow";
