// @base-ui-components/react@1.0.0-beta.0/esm/preview-card/trigger/PreviewCardTrigger.js
'use client';

import * as React from 'react';
import { useComponentRenderer } from "../../utils/useComponentRenderer.js";
import { usePreviewCardRootContext } from "../root/PreviewCardContext.js";
import { useForkRef } from "../../utils/useForkRef.js";
import { triggerOpenStateMapping } from "../../utils/popupStateMapping.js";

/**
 * A link that opens the preview card.
 * Renders an `<a>` element.
 *
 * Documentation: [Base UI Preview Card](https://base-ui.com/react/components/preview-card)
 */
export const PreviewCardTrigger = /*#__PURE__*/React.forwardRef(function PreviewCardTrigger(props, forwardedRef) {
  const {
    render,
    className,
    ...otherProps
  } = props;
  const {
    open,
    getRootTriggerProps,
    setTriggerElement
  } = usePreviewCardRootContext();
  const state = React.useMemo(() => ({
    open
  }), [open]);
  const mergedRef = useForkRef(setTriggerElement, forwardedRef);
  const {
    renderElement
  } = useComponentRenderer({
    propGetter: getRootTriggerProps,
    render: render ?? 'a',
    className,
    state,
    ref: mergedRef,
    extraProps: otherProps,
    customStyleHookMapping: triggerOpenStateMapping
  });
  return renderElement();
});
if (process.env.NODE_ENV !== "production") PreviewCardTrigger.displayName = "PreviewCardTrigger";
