// @base-ui-components/react@1.0.0-beta.0/cjs/radio/indicator/RadioIndicator.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RadioIndicator = void 0;
var React = _interopRequireWildcard(require("react"));
var _useComponentRenderer = require("../../utils/useComponentRenderer");
var _RadioRootContext = require("../root/RadioRootContext");
var _customStyleHookMapping = require("../utils/customStyleHookMapping");
var _useOpenChangeComplete = require("../../utils/useOpenChangeComplete");
var _useForkRef = require("../../utils/useForkRef");
var _useTransitionStatus = require("../../utils/useTransitionStatus");
/**
 * Indicates whether the radio button is selected.
 * Renders a `<span>` element.
 *
 * Documentation: [Base UI Radio](https://base-ui.com/react/components/radio)
 */
const RadioIndicator = exports.RadioIndicator = /*#__PURE__*/React.forwardRef(function RadioIndicator(props, forwardedRef) {
  const {
    render,
    className,
    keepMounted = false,
    ...otherProps
  } = props;
  const rootState = (0, _RadioRootContext.useRadioRootContext)();
  const rendered = rootState.checked;
  const {
    transitionStatus,
    setMounted
  } = (0, _useTransitionStatus.useTransitionStatus)(rendered);
  const state = React.useMemo(() => ({
    ...rootState,
    transitionStatus
  }), [rootState, transitionStatus]);
  const indicatorRef = React.useRef(null);
  const mergedRef = (0, _useForkRef.useForkRef)(forwardedRef, indicatorRef);
  const {
    renderElement
  } = (0, _useComponentRenderer.useComponentRenderer)({
    render: render ?? 'span',
    ref: mergedRef,
    className,
    state,
    extraProps: otherProps,
    customStyleHookMapping: _customStyleHookMapping.customStyleHookMapping
  });
  (0, _useOpenChangeComplete.useOpenChangeComplete)({
    open: rendered,
    ref: indicatorRef,
    onComplete() {
      if (!rendered) {
        setMounted(false);
      }
    }
  });
  const shouldRender = keepMounted || rendered;
  if (!shouldRender) {
    return null;
  }
  return renderElement();
});
if (process.env.NODE_ENV !== "production") RadioIndicator.displayName = "RadioIndicator";
